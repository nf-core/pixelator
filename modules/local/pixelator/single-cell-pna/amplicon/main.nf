process PIXELATOR_PNA_AMPLICON {
    tag "${meta.id}"
    label 'process_medium'

    // TODO: Add conda
    // conda "bioconda::pixelator=0.18.2"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(reads, arity: '1..*')

    output:
    tuple val(meta), path("amplicon/*.amplicon.{fq,fastq}.zst"), emit: amplicon
    tuple val(meta), path("amplicon/*.report.json"),             emit: report_json
    tuple val(meta), path("amplicon/*.meta.json"),               emit: metadata_json
    tuple val(meta), path("*pixelator-amplicon.log"),            emit: log

    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''
    def r = reads

    // Make list of old name and new name pairs to use for renaming
    // Use R1/R2 style suffixes for limited backward compatibility with pixelator<0.17
    def old_new_pairs = (reads.size() == 1
        ? [[reads[0], "${prefix}${getFileSuffix(reads[0])}"]]
        : reads.withIndex().collect { entry, index -> [entry, "${prefix}_R${index + 1}${getFileSuffix(entry)}"] })
    // Flatten a list of tuples into a single string joined with spaces
    def rename_to = old_new_pairs.flatten().join(' ')
    def renamed_reads = old_new_pairs.collect { old_name, new_name -> new_name }.join(' ')


    """
    printf "%s %s\\n" ${rename_to} | while read old_name new_name; do
        [ -f "\${new_name}" ] || ln -s \$old_name \$new_name
    done

    pixelator \\
        --log-file ${prefix}.pixelator-amplicon.log \\
        --verbose \\
        single-cell-pna \\
        amplicon \\
        --threads ${task.cpus} \\
        --output . \\
        ${args} \\
        ${renamed_reads}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir amplicon
    touch amplicon/${prefix}.report.json
    touch amplicon/${prefix}.meta.json
    touch amplicon/${prefix}.amplicon.fq.zst
    touch ${prefix}.pixelator-amplicon.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}



// for .gz files also include the second to last extension if it is present. E.g., .fasta.gz
// Source: nf-core/modules/cat/cat
def getFileSuffix(filename) {
    def match = filename =~ /^.*?((\.\w{1,5})?(\.\w{1,5}\.(gz|zst)$))/
    return match ? match[0][1] : filename.substring(filename.lastIndexOf('.'))
}
