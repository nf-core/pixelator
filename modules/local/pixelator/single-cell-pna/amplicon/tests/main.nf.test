nextflow_process {

    name "Test Process PIXELATOR_PNA_AMPLICON"
    script "../main.nf"
    process "PIXELATOR_PNA_AMPLICON"
    tag "modules"
    tag "pixelator"
    tag "pixelator/pna"
    tag "pixelator/single_cell_pna_amplicon"

    test("PNA amplicon - small test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/pna/fastq/PNA055_Sample07_filtered_S7_R1_001.fastq.gz', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/pna/fastq/PNA055_Sample07_filtered_S7_R2_001.fastq.gz', checkIfExists: true)
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.amplicon).match("amplicon") },
                { assert snapshot(process.out.report_json).match("report_json") },
                { assert snapshot(process.out.metadata_json).match("metadata_json") },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator amplicon")
                    }
                },
            )
        }
    }
}

