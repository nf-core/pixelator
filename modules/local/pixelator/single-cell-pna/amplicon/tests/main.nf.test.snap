{
    "amplicon": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.amplicon.fq.zst:md5,460f0dd9af7c5ebec8bcad76e3964b7d"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-27T08:26:22.987687656"
    },
    "report_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.report.json:md5,ae3608057ebf0eceaae5b3c0efa8320b"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-22T17:25:27.100151311"
    },
    "metadata_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.meta.json:md5,59094765101eb355c10938eff88a24e1"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-22T17:25:27.236295681"
    }
}